### ISSUE TITLE
Why doesn’t kl_div ignore -100 in pseudo labels?

### ISSUE THREAD

**Issue by yynil** [jxiw/MambaInLlama#11]
The original codes looks like below:
            kl_loss = F.kl_div(F.log_softmax(
                student_logits, dim=-1), targets, reduction='batchmean')

Although the relative loss curve is the same, the optimizer might try to optimize the unnecessary targets. I think it will be better to utilize the -100 indexes as mask to inputs.

**Comment by jxiw** [comment]
Thank you for your insightful observation. It absolutely makes sense. You're right that the optimizer might end up focusing on unnecessary targets, which isn't ideal. I haven't fully optimized this part yet, and using the -100 indexes as a mask to filter out irrelevant inputs would definitely be a more efficient approach. Thank you for your suggestion!

**Comment by yynil** [comment]
actually I doubt if this modification can  eliminate the needs to do SFT/DPO. I made a modified version to distill a RWKV in Llama3.1. Let’s see what will happen in this modified loss calculation.

**Comment by yynil** [comment]
Thanks for your kindly help, that's really very important and insightful for me! 
I also have another question, I tried a very aggressive distillation that distill all 32 layers of LLama in one hybrid training process, but I got a very low student's cross entropy loss and a very high and noisy kl loss no matter I only calculate label's kl or all kl loss. Could you share your experience about the kl-div loss? If the problem lies in the aggressive distillation(32 layers together) ?

**Comment by jxiw** [comment]
Regarding the KL loss, my best model with 1/8 attention achieves under 50. I am still trying to train it with more data to push it further, but I have limited GPU resources.

Here are some suggestions based on my experience:

1) If you are only interested in non-attention models (and don't want to use any attention), I suggest distilling from the first layer to the second, and then along this way, to the last layer. The stepwise distillation definitely improves form my experience.

2) You may want to consider switching to the Llama3 model instead of Llama3 instructed, as Llama3 instructed uses RLHF for alignment, which is still a bit of a mystery for us. So, you could consider using general corpora like RefineWeb and using Llama3 as the teacher model. 

3) You may want to consider adding MSE loss for the hidden output of each layer, similar to what's described in[this](https://arxiv.org/pdf/2010.13002).  But to be honest, from my experiment using stepwise distillation, that does not happy a lot.

**Comment by yynil** [comment]
Really insightful and directional instructions! I tried to distill Instruction Llama3.1 with all labels kl div in the first layer, the common sense dialogue seems as good as the original llama.  In some special case like comparing 9.11 and 9.8, the hybrid model is even better. I will continue to distill and SFT the rest steps to see what will happen then.
Thanks again!

**Comment by jxiw** [comment]
We are glad to hear that! If you have any further questions, please feel free to reach out at any time.
